### PR TITLE
Add feature to display logs on UI 

### PR DESCRIPTION
**Pull request by KhalidRmb**
@Ram81 @RishabhJain2018 @deshraj

### CODE REVIEW

**Comment by KhalidRmb** [comment]
Will fix the tests. Failing due to change in logging messages.

File: apps/challenges/aws_utils.py [hunk 1]
```diff
@@ -185,7 +186,7 @@
             "logConfiguration": {{
                 "logDriver": "awslogs",
                 "options": {{
-                    "awslogs-group": "evalai-worker-{ENV}",
+                    "awslogs-group": "challenge-pk-{challenge_pk}-workers",
```
Comment (Ram81) [inline apps/challenges/aws_utils.py:189]: @KhalidRmb how will this affect existing worker logs? they'll be in separate log group right?

File: apps/challenges/views.py [hunk 2]
```diff
@@ -2553,3 +2553,35 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+def get_worker_logs(request, challenge_pk):
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized to access the worker logs."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    challenge = challenge = Challenge.objects.get(pk=challenge_pk)
+    response_data = []
+    last_submission = Submission.objects.order_by('started_at').last()
```
Comment (Ram81) [inline apps/challenges/views.py:2571]: What about the case when some intermediate submissions failed? In that case API won't send any logs?

Comment (KhalidRmb) [reply apps/challenges/views.py:2571]: I actually need to update the code to check for submissions of only the current challenge. Will push the change in next commit.

Regarding the case where some intermediate submission fails: I remember Rishabh saying once a submission passes all the future submissions for the challenge would pass. That's the reason I was checking only for the latest submission.

File: apps/challenges/aws_utils.py (hunk 1 above)
Comment (KhalidRmb) [reply apps/challenges/aws_utils.py:189]: Those logs would still be accessible on CloudWatch I think. The display will only work for the newly launched workers from here on.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 3]
```diff
@@ -681,7 +739,7 @@
             parameters.callback = {
                 onSuccess: function(response) {
                     var details = response.data;
-                    vm.leaderboard = details.results;
+                    vm.leaderboard = details.results; console.log("Just updated leaderboard as well.")
```
Comment (Ram81) [inline frontend/src/js/controllers/challengeCtrl.js:742]: remove the console statement

File: apps/challenges/views.py [hunk 4]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+    elif action == 'restart':
+        response = restart_workers([challenge])
+        count, failures = response["count"], response["failures"]
+        if count:
+            response_data = {"action": "Success"}
+        else:
+            message = failures[0]["message"]
+            response_data = {"action": "Failure", "error": message}
+
+    return Response(response_data, status=status.HTTP_200_OK)
```
Comment (Ram81) [inline apps/challenges/views.py:2631]: add newline

File: apps/challenges/views.py [hunk 5]
```diff
@@ -110,6 +111,7 @@
     ZipChallengeSerializer,
     ZipChallengePhaseSplitSerializer,
 )
+from.aws_utils import start_workers, stop_workers, restart_workers, get_logs_from_cloudwatch
```
Comment (Ram81) [inline apps/challenges/views.py:114]: why is there no space between `from` and `.aws_utils` ?

**Comment by RishabhJain2018** [comment]
@KhalidRmb Please resolve the conflicts here.

File: apps/challenges/urls.py [hunk 6]
```diff
@@ -166,4 +166,14 @@
         views.validate_challenge_config,
         name="validate_challenge_config",
     ),
+    url(
+        r"^challenge/(?P<challenge_pk>[0-9]+)/get_worker_logs/$",
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:170]: `r"^challenge/(?P<challenge_pk>[0-9]+)/get_worker_logs/$",` -> `r"^(?P<challenge_pk>[0-9]+)/get_worker_logs/$",`

File: apps/challenges/urls.py [hunk 7]
```diff
@@ -166,4 +166,14 @@
...
+        views.get_worker_logs,
+        name="get_worker_logs"
+    ),
+    url(
+        r"^challenge/(?P<challenge_pk>[0-9]+)/manage_worker/(?P<action>[\w-]+)$",
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:175]: `r"^challenge/(?P<challenge_pk>[0-9]+)/manage_worker/(?P<action>[\w-]+)$",`-> `r"^(?P<challenge_pk>[0-9]+)/manage_worker/(?P<action>[\w-]+)$",`

File: apps/challenges/views.py [hunk 8]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((ExpiringTokenAuthentication,))
+def get_worker_logs(request, challenge_pk):
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized to access the worker logs."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    challenge = challenge = Challenge.objects.get(pk=challenge_pk)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2571]: Can we please use the `get_challenge_model` function here.

File: apps/challenges/views.py [hunk 9]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+    response_data = []
+
+    log_group_name = "challenge-pk-{}-workers".format(challenge_pk)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2574]: Let's use `challenge.pk` here.

File: apps/challenges/views.py [hunk 10]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+
+
+@api_view(["PUT"])
+@throttle_classes([UserRateThrottle])
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((ExpiringTokenAuthentication,))
+def manage_worker(request, challenge_pk, action):
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized."
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2597]: ` "error": "Sorry, you are not authorized for access worker operations."`

File: apps/challenges/views.py [hunk 11]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    challenge = Challenge.objects.get(pk=challenge_pk)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2601]: Can we please use the `get_challenge_model` function here.

File: apps/challenges/views.py [hunk 12]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+
+    response_data = {}
+
+    if action == 'start':
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2605]: Can we please be consistent in quoting. We follow double quotes everywhere.

File: apps/challenges/views.py [hunk 13]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+        response = start_workers([challenge])
+        count, failures = response["count"], response["failures"]
+        logging.info("Count is {} and failures are: {}".format(count, failures))
+        if count:
+            response_data = {"action": "Success"}
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2610]: `response_data = {"action": "Worker(s) is successfully started"}`

File: apps/challenges/views.py [hunk 14]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+        else:
+            message = failures[0]["message"]
+            response_data = {"action": "Failure", "error": message}
+    elif action == 'stop':
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2614]: Can we please be consistent in quoting. We follow double quotes everywhere.

File: apps/challenges/views.py [hunk 15]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+        response = stop_workers([challenge])
+        count, failures = response["count"], response["failures"]
+        if count:
+            response_data = {"action": "Success"}
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2618]: `response_data = {"action": "Worker(s) is successfully stopped"}`

File: apps/challenges/views.py [hunk 16]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+        else:
+            message = failures[0]["message"]
+            response_data = {"action": "Failure", "error": message}
+    elif action == 'restart':
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2622]: Can we please be consistent in quoting. We follow double quotes everywhere.

File: apps/challenges/views.py [hunk 17]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+        response = restart_workers([challenge])
+        count, failures = response["count"], response["failures"]
+        if count:
+            response_data = {"action": "Success"}
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2626]: response_data = {"action": "Worker(s) is successfully restarted"}

File: frontend/src/js/controllers/challengeCtrl.js [hunk 18]
```diff
@@ -74,6 +75,8 @@
         vm.isChallengeLeaderboardPrivate = false;
         vm.previousPublicSubmissionId = null;
 
+        vm.worker_logs = []
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js:78]: Can we please use CamelCase here.

**Comment by RishabhJain2018** [comment]
Also, @KhalidRmb I feel when you commit code it is not autoformatted by black. Can you please check if you have the hook installed in your repository.

File: apps/challenges/views.py (hunk 17 above)
Comment (KhalidRmb) [reply apps/challenges/views.py:2626]: Handling it dynamically in the frontend. This is just to denote success.

File: apps/challenges/views.py (hunk 15 above)
Comment (KhalidRmb) [reply apps/challenges/views.py:2618]: Handling the message dynamically in the frontend. This is just to denote the success of whatever action was chosen.

File: apps/challenges/views.py (hunk 13 above)
Comment (KhalidRmb) [reply apps/challenges/views.py:2610]: Handling the message dynamically in the frontend. This is just to denote the success of whatever action was chosen.

Comment (RishabhJain2018) [reply apps/challenges/views.py:2610]: okay!

File: apps/challenges/urls.py [hunk 19]
```diff
@@ -166,4 +166,14 @@
         views.validate_challenge_config,
         name="validate_challenge_config",
     ),
+    url(
+        r"^(?P<challenge_pk>[0-9]+)/get_worker_logs/$",
+        views.get_worker_logs,
+        name="get_worker_logs"
+    ),
+    url(
+        r"^(?P<challenge_pk>[0-9]+)/manage_worker/(?P<action>[\w-]+)$",
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:175]: `r"^(?P<challenge_pk>[0-9]+)/manage_worker/(?P<action>[\w-]+)$",` -> `r"^(?P<challenge_pk>[0-9]+)/manage_worker/(?P<action>[\w-]+)/$"`,

Comment (RishabhJain2018) [reply apps/challenges/urls.py:175]: Forward slash missing at the end of URL.

File: apps/challenges/views.py [hunk 20]
```diff
@@ -2553,3 +2555,77 @@ def validate_challenge_config(request, challenge_host_team_pk):
...
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((ExpiringTokenAuthentication,))
+def get_worker_logs(request, challenge_pk):
+    if not is_user_a_host_of_challenge(request.user, challenge_pk):
+        response_data = {
+            "error": "Sorry, you are not authorized to access the worker logs."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    challenge = Challenge.objects.get(pk=challenge_pk)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2571]: Can we please use `challenge = get_challenge_model(challenge_pk)`  here.

**Pull request merged**